plugins/net_plugin/net_plugin.cpp

```diff
@@ -4333,47 +4443,105 @@ namespace eosio {
    // called by API
    string connections_manager::connect( const string& host, const string& p2p_address ) {
+      std::unique_lock g( connections_mtx );
+      supplied_peers.insert(host);
+      g.unlock();
+      return resolve_and_connect( host, p2p_address );
+   }
+
+   string connections_manager::resolve_and_connect( const string& peer_address, const string& listen_address ) {
+      string::size_type colon = peer_address.find(':');
+      if (colon == std::string::npos || colon == 0) {
+         fc_elog( logger, "Invalid peer address. must be \"host:port[:<blk>|<trx>]\": ${p}", ("p", peer_address) );
+         return "invalid peer address";
+      }
+
       std::lock_guard g( connections_mtx );
-      if( find_connection_i( host ) )
+      if( find_connection_i( peer_address ) )
          return "already connected";
 
-      connect_i( host, p2p_address );
-      supplied_peers.insert(host);
+      auto [host, port, type] = split_host_port_type(peer_address);
+
+      auto resolver = std::make_shared<tcp::resolver>( my_impl->thread_pool.get_executor() );
+
+      resolver->async_resolve(host, port, 
+         [resolver, host = host, port = port, peer_address = peer_address, listen_address = listen_address, this]( const boost::system::error_code& err, const tcp::resolver::results_type& results ) {
+            connection_ptr c = std::make_shared<connection>( peer_address, listen_address );
+            c->set_heartbeat_timeout( heartbeat_timeout );
+            std::lock_guard g( connections_mtx );
+            auto [it, inserted] = connections.emplace( connection_details{
+               .host = peer_address,
+               .c = std::move(c),
+               .ips = results
+            });
+            if( !err ) {
+               it->c->connect( results, it );
+            } else {
+               fc_elog( logger, "Unable to resolve ${host}:${port} ${error}",
+                        ("host", host)("port", port)( "error", err.message() ) );
+               it->c->set_state(connection::connection_state::closed);
+               ++(it->c->consecutive_immediate_connection_close);
+            }
+      } );
+
       return "added connection";
    }
 
+   void connections_manager::update_connection_endpoint(connection_details_index::const_iterator it,
+                                                        const tcp::endpoint& endpoint) {
+      fc_dlog(logger, "updating connection endpoint");
+      std::unique_lock g( connections_mtx );
+      auto& index = connections.get<by_active_ip>();
+      index.modify_key(connections.project<by_active_ip>(it), [endpoint](tcp::endpoint& e) {
+         e = endpoint;
+      });
+   }
+
+   connections_manager::connection_details_index::iterator connections_manager::get_connection_iterator(const connection_ptr& c) {
+      std::lock_guard g( connections_mtx );
+      const auto& index = connections.get<by_connection>();
+      auto i = index.find(c);
+      return connections.project<by_host>(i);
```

Comment:
The lock is released here. The returned iterator can't be trusted after the lock is released.

Reply:
Call chain reworked to not pass an iterator.